services/121-service/src/payments/fsp-integration/intersolve-visa/intersolve-visa.api.service.ts

```diff
@@ -119,52 +225,219 @@ export class IntersolveVisaApiService {
+    tokenCode: string,
+  ): Promise<GetTransactionInformationReturnType> {
+    // get Transactions
+    const getTransactionsResponseDto = await this.getTransactions({
+      tokenCode,
+      fromDate: this.getTwoMonthsAgo(),
+    });
+
+    // Seperate out the reservation and expired reservation transactions.
+    const transactions = getTransactionsResponseDto.data.data;
+    let reservationTransactions: IntersolveGetTransactionsResponseDataDto[] =
+      [];
+    let expiredTransactions: IntersolveGetTransactionsResponseDataDto[] = [];
+    if (transactions && transactions.length > 0) {
+      reservationTransactions = transactions.filter(
+        (t) => t.type === 'RESERVATION',
+      );
+      expiredTransactions = transactions.filter(
+        (t) => t.type === 'RESERVATION_EXPIRED',
+      );
+    }
+
+    // Determine the last used date of the reservation transactions
+    const lastTransactionDate = this.getLastTransactionDate(
+      reservationTransactions,
     );
+
+    // Calculate the amount spent this month from the reservation and expired transactions
+    const spentThisMonth = this.calculateSpentThisMonth({
+      walletTransactions: reservationTransactions,
+      expiredReserveTransactions: expiredTransactions,
+    });
+
+    // Return relevant information
+    const getTransactionInformationResultDto: GetTransactionInformationReturnType =
+      {
+        spentThisMonth: spentThisMonth,
+        lastTransactionDate: lastTransactionDate,
+      };
+    return getTransactionInformationResultDto;
   }
 
-  public async getTransactions(
-    tokenCode: string | null,
-    dateFrom?: Date,
-  ): Promise<GetTransactionsDetailsResponseDto> {
+  private async getTransactions({
+    tokenCode,
+    fromDate,
+  }: {
+    tokenCode: string;
+    fromDate?: Date;
+  }): Promise<GetTransactionsResponseDto> {
+    // Send the request
     const authToken = await this.getAuthenticationToken();
     const apiPath = process.env.INTERSOLVE_VISA_PROD
       ? 'wallet-payments'
       : 'wallet';
     const url = `${intersolveVisaApiUrl}/${apiPath}/v1/tokens/${tokenCode}/transactions${
-      dateFrom ? `?dateFrom=${dateFrom.toISOString()}` : ''
+      fromDate ? `?dateFrom=${fromDate.toISOString()}` : ''
     }`;
     const headers = [
       { name: 'Authorization', value: `Bearer ${authToken}` },
       { name: 'Tenant-ID', value: process.env.INTERSOLVE_VISA_TENANT_ID },
     ];
-    return await this.httpService.get<any>(url, headers);
+
+    const getTransactionsResponseDto =
+      await this.httpService.get<GetTransactionsResponseDto>(url, headers);
+    // Handle the response
+
+    const errorMessage = this.createErrorMessageIfRequestFailed(
+      getTransactionsResponseDto,
+    );
+    // If the response contains errors
+    if (errorMessage) {
+      throw new IntersolveVisaApiError(
+        `${IntersolveVisa121ErrorText.getTransactionError}: ${errorMessage}`,
+      );
+    }
+
+    return getTransactionsResponseDto;
```

Comment:
I thought about this, but decided against it, because I wanted to throw errors at the place where something "went wrong", and not in a helper method. Because then the stack trace etc. you get is from the actual place, and not all coming from that helper method.

Reply:
The stack trace would still include the calling function, or in other words, the function calling `httpRequestWithErrorHandling`. I don't think that's a relevant reason not to do this.

Reply:
> If this is a pattern that keeps coming back in several services, then this could be moved to a base class that the services extend from.

Or introduce a utility/helper class?

Reply:
Let's pair program the solution, I am unfamiliar with this construct.

Reply:
Yep, happy to pair-program it. Utility/helper class could also work - we can explore which might be better suited here together.

Reply:
TODO: Check out IntersolveVisa BaseREsoinseDTO